commit 9c6e71079b46ad5433165feaa2001450f2017b56
Author: Przemysław Buczkowski <prem@prem.moe>
Date:   Mon Aug 16 13:16:21 2021 +0100

    GCC: Patch for Apple Silicon compatibility
    
    This patch fixes a linker error occuring when compiling
    the cross-compiler on macOS and ARM64 architecture.
    
    Adapted from:
    https://github.com/richfelker/musl-cross-make/issues/116#issuecomment-823612404
    
    Change-Id: Ia3ee98a163bbb62689f42e2da83a5ef36beb0913
    Reviewed-on: https://review.haiku-os.org/c/buildtools/+/4329
    Reviewed-by: John Scipione <jscipione@gmail.com>
    Reviewed-by: Adrien Destugues <pulkomandy@gmail.com>

--- a/gcc/config/aarch64/aarch64.h
+++ b/gcc/config/aarch64/aarch64.h
@@ -1293,7 +1293,7 @@ extern const char *aarch64_rewrite_mcpu
 #define MCPU_TO_MARCH_SPEC_FUNCTIONS \
   { "rewrite_mcpu", aarch64_rewrite_mcpu },
 
-#if defined(__aarch64__)
+#if defined(__aarch64__) && ! defined(__APPLE__)
 extern const char *host_detect_local_cpu (int argc, const char **argv);
 #define HAVE_LOCAL_CPU_DETECT
 # define EXTRA_SPEC_FUNCTIONS						\
--- a/gcc/config/host-darwin.cc
+++ b/gcc/config/host-darwin.cc
@@ -23,6 +23,8 @@
 #include "options.h"
 #include "diagnostic-core.h"
 #include "config/host-darwin.h"
+#include "hosthooks.h"
+#include "hosthooks-def.h"
 #include <errno.h>
 
 /* For Darwin (macOS only) platforms, without ASLR (PIE) enabled on the
@@ -181,3 +183,5 @@ darwin_gt_pch_use_address (void *&addr,
 
   return 1;
 }
+
+const struct host_hooks host_hooks = HOST_HOOKS_INITIALIZER;
--- a/gcc/system.h
+++ b/gcc/system.h
@@ -197,2 +197,2 @@
-/* Include <string> before "safe-ctype.h" to avoid GCC poisoning
-   the ctype macros through safe-ctype.h */
+/* Include requested C++ headers before "safe-ctype.h" to avoid GCC poisoning
+   the ctype macros through safe-ctype.h.  */
@@ -200,0 +201,21 @@
+#if defined (INCLUDE_ALGORITHM) || !defined (HAVE_SWAP_IN_UTILITY)
+# include <algorithm>
+#endif
+#ifdef INCLUDE_LIST
+# include <list>
+#endif
+#ifdef INCLUDE_MAP
+# include <map>
+#endif
+#ifdef INCLUDE_SET
+# include <set>
+#endif
+#ifdef INCLUDE_VECTOR
+# include <vector>
+#endif
+#ifdef INCLUDE_ARRAY
+# include <array>
+#endif
+#ifdef INCLUDE_FUNCTIONAL
+# include <functional>
+#endif
@@ -220,21 +241,0 @@
-#if defined (INCLUDE_ALGORITHM) || !defined (HAVE_SWAP_IN_UTILITY)
-# include <algorithm>
-#endif
-#ifdef INCLUDE_LIST
-# include <list>
-#endif
-#ifdef INCLUDE_MAP
-# include <map>
-#endif
-#ifdef INCLUDE_SET
-# include <set>
-#endif
-#ifdef INCLUDE_VECTOR
-# include <vector>
-#endif
-#ifdef INCLUDE_ARRAY
-# include <array>
-#endif
-#ifdef INCLUDE_FUNCTIONAL
-# include <functional>
-#endif
--- a/libcc1/libcc1plugin.cc
+++ b/libcc1/libcc1plugin.cc
@@ -33,0 +34,2 @@
+#include <vector>
+
@@ -71,2 +72,0 @@
-
-#include <vector>
--- a/libcc1/libcp1plugin.cc
+++ b/libcc1/libcp1plugin.cc
@@ -34,0 +35,2 @@
+#include <vector>
+
@@ -73,2 +74,0 @@
-
-#include <vector>
